src/dataflow/src/render/reduce.rs

```diff
@@ -1413,6 +1547,41 @@ where
                             (AggregateFunc::SumDecimal, AccumInner::SimpleNumber { accum, .. }) => {
                                 Datum::from(*accum)
                             }
+                            (
+                                AggregateFunc::SumAPD,
+                                AccumInner::APD {
+                                    accum,
+                                    pos_infs,
+                                    neg_infs,
+                                    nans,
+                                    non_nulls,
+                                },
+                            ) => {
+                                let mut cx_datum = apd::cx_datum();
+                                let d = cx_datum.to_width(accum.0);
+                                // Take a wide decimal (aggregator) into a
+                                // narrow decimal (datum). If this operation
+                                // overflows the datum, this new value will be
+                                // +/- infinity. However, the aggregator tracks
+                                // the amount of overflow, making it invertible.
```

Comment:
I don't understand this sentence. The `reduce` code is pretty agnostic about what you do, except that it wants to be sure that the accumulation is associative and commutative. It doesn't look like the code that guarantees that is in this file, unless these really are *arbitrary* precision decimals, but I thought at some point this wasn't actually the case. It now occurs to me that they are called `apd`, but if so where does the risk of overflow come from?

Reply:
I believe we're preserving associativity and commutativity in the aggregated values in this file. Glad to offer a brief explanation and if I am mistaken about the invariants we need to offer, glad to chat more at length about what is missing.

We preserve associativity and commutativity in the accumulated valued (I call this an aggregator in some places, which is language I should maybe tighten up), which is an 81-digit-precision value. The values that we allow to modify this aggregated value are all input as 39-digit-precision values, i.e. the datums we'll receive from clients. This guarantee that the input is so much narrower lets us accept a lot of values that max out the narrower preicison. i.e. if we receive `1e38` and `1e-39`, the aggregator can store their sum without needing to round. The moment we receive 39-digit input that causes an aggregator to exeed 81 digits of precison, we panic ([1](https://github.com/MaterializeInc/materialize/pull/6872/files/20fcfe3bf31349d550bf6e2f97cd3fa3b22587f3#diff-1d8a1661e2a84d234a135ddf4fca18e32c15394674892cb74cbc877a7d91d36eR1158), [2](https://github.com/MaterializeInc/materialize/pull/6872/files/20fcfe3bf31349d550bf6e2f97cd3fa3b22587f3#diff-1d8a1661e2a84d234a135ddf4fca18e32c15394674892cb74cbc877a7d91d36eR1218)), which was the design we had agreed on [a while back](https://github.com/MaterializeInc/materialize/pull/6283#issuecomment-812049649).

Just for my own sanity, here's a high-level overview of the process.

## Input

1. We receive datums (39-digit-precision values), and take them to [81-digit-precision values](https://github.com/MaterializeInc/materialize/pull/6872/files/20fcfe3bf31349d550bf6e2f97cd3fa3b22587f3#diff-1d8a1661e2a84d234a135ddf4fca18e32c15394674892cb74cbc877a7d91d36eR1365-R1368).
2. These widened datums [are added to the existing aggregated value](https://github.com/MaterializeInc/materialize/pull/6872/files/20fcfe3bf31349d550bf6e2f97cd3fa3b22587f3#diff-1d8a1661e2a84d234a135ddf4fca18e32c15394674892cb74cbc877a7d91d36eR1141-R1157), wherein we ensure that we haven't exceeded that value's max precision of 81 digits.

## Output

When we return the aggregated value, we take the wide type (an aggregator) into a narrow type, which happens near this comment. Because the wide value can exceed what the narrow value can hold, we allow it to signal an overflow by generating infinity. However, this doesn't modify the aggregated value. This way, if we receive a retraction for the data that caused the overflow, we can subtract it, moving us back into non-overflow territory, and we can return literal values once again.

I'm sure this is second explanation is either bad or unncessary, but trying to make what's happening clear so I can make whatever adjustments we need to make sure whoever comes across this next isn't puzzled.

You can see this process in flight in the [`slt`](https://github.com/MaterializeInc/materialize/pull/6872/files/20fcfe3bf31349d550bf6e2f97cd3fa3b22587f3#diff-feb8bc7e759d4edf7427cddf44b9387d0bb7d6c71224e99fabedf94883507a3cR73-R97). I am working on making these tests more faithfully retract values rather than just perform a negative addition, but am blocked on [some fussy details](https://github.com/MaterializeInc/rust-dec/pull/35) in the upstream library, which renders `DISTINCT` inoperable due to passing around some uninitialized values.

Reply:
Ok, I think some of what I was missing is in https://github.com/MaterializeInc/materialize/pull/6872/files#diff-1d8a1661e2a84d234a135ddf4fca18e32c15394674892cb74cbc877a7d91d36eR1156-R1158 which it seems like it must be the code that prevents the system from surviving incorrect math. I suspect it is clear to you, but def not to me, what the guarantees of this fragment are (probably `rounded()` does something; is it true whenever there is loss of information?).

Also, just as a sanity check: the use of `APD` and `apd` all over the place seems not exactly right, if the decimals are not arbitrary precision. Am I being too pedantic there?

Reply:
Ah, `rounded` has a very specific meaning in the library, which [I'll include links to the docs for](http://speleotrove.com/decimal/dncont.html), and sorry I hadn't already!

>  The DEC_Rounded flag is set whenever a result or input operand is rounded (even if only zero digits were discarded). 

This is essentially the most sensitive tooling available to show that a value needed to be rounded, which only happens when you've exceeded its max precision.

As for the name `apd`, definitely ill-chosen, but also temporary because this will become `decimal`. If it's a real nuisance, glad to commit a rename/